### ISSUE TITLE
CLI: Enable better tests and more use cases; avoid env vars in tests

### ISSUE THREAD

**Issue by Enrico2** [apollographql/rust#67]
The CLI uses some environment variables to get certain values at runtime.

This makes some functionality hard to test in parallel (in both unit tests and integration tests), because changing an environment variable in one thread, affects other threads.
Most recently #65 fixed some flaky tests that experienced this problem from `APOLLO_CDN_URL`.

A straw dog proposal for now, is to use thread-local values as an override for environment variables, i.e. when trying to get a value for X, first check thread-local for X, and if it's not there, try to get it from the environment.

This is currently blocking writing unit tests for anything that uses `APOLLO_CDN_URL` in the tested code.

**Comment by jbaxleyiii** [comment]
@Enrico2 yeah this a big problem of my inital path of doing things 👎 We do want people to be able to use env variables to override parts of the CLI so I was wondering if investing in a way to mock those better from a testing perspective might be the right way to go?

**Comment by Enrico2** [comment]
@jbaxleyiii Can you elaborate on why we'd want environment variables and not command line flags? I've really seen the latter used a lot more than the former.

(Regardless, I think we can set it up such that the **binary** is configurable with environment variables, but it immediately moves the data from the environment variables to something else; a struct, thread local values, etc. Then tests would mock/create the values there directly.)

**Comment by ashleygwilliams** [comment]
i'm asked in the design doc (not *yet* on github ;)) about what our vision and guiding values are around balancing flags, env vars, and config. there's a lot of potential options here. admittedly i try very hard to reduce the number of flags my CLI has, preferring config with env vars to override. especially in the universe where we are doing this for *ourselves* for our own tests, cluttering up the interface with flags that folks would need to use as a one off or for rare/uncommon usecases, feels like the wrong strategy.

we could continue the convo here, but i do think coming up with an overall strategy first would be a great first step to addressing this!

**Comment by Enrico2** [comment]
> cluttering up the interface with flags that folks would need to use as a one off or for rare/uncommon usecases, feels like the wrong strategy.

Yes I wouldn't want them to show up in `--help`, we can make those flags hidden though. 

Is there a good solution for tests to run in parallel that might depend on different values of something that we define globally?

**Comment by jbaxleyiii** [comment]
@Enrico2 I think this is a great discusion that we should reflect in the CLI UX doc (which we should move to this repo nowish).

My basic view of the CLI design for flags / config / env vars breaks down like this:

- Flags: flags are scoped to a single execution of a command, act as named arguments much like you for an API call in code (and preferred over multiple positional arguments), and are often scoped to a command subset (i.e. all `ap schema` commands may used a `--graph` flag
- Config: used for project / machine stable values that persist for the lifetime of the project. These shouldn't be secrets, shouldn't change based on environment, and should be the same across execution of all commands and users of the tool. Global tool settings, project settings, etc live here for a tool that isn't interactive (i.e an editor) and prefer to not have config for commands unless the information isn't part of the command itself. An example of that is identity of the project. Everytime I go to call the registry I shouldn't need to specify who I am in relationship to the project if I can set it in a stable project config (i.e. a name)
- env variables: These are values that should persist across commands but be specific to the environment in which the command is being run. That environment may be *person* or the *machine* but they aren't stable values (i.e. they can be volatile / generated), secrets, or properties owned by the environment itself. An example of this may be the registry url. This should persist across every invocation of the CLI in a given CI job but isn't something that every invocation of the CLI in that repostitory may need to use. Or I may want to set it as my env for a period of time while testing against staging or using direnv within a companies project, etc.

Ultimately these different types of controls will need to layer as they will overlap and we need to work together as a team to understand when to go to each. Looping back to the original question of the issue, I do think that we should be able to support using the env as a command override surface area in test suites as a general statement and specfically for the registry url as it relates to the current bottle neck. I also don't know how to do this with rust!

**Comment by Enrico2** [comment]
I really like that we clearly define the criteria, but I'm not sure about the specifics.
Most of this is great, and makes a lot of sense to me, but there are some exceptions we should consider.

> **Env variables**: These are values that should persist across commands but be specific to the environment in which the command is being run.

_My_ interpretation of this sentence along with 

> **Flags**: flags are scoped to a single execution of a command

Makes me thing that something like `endpoint-url` should be both an environment variable, as it should persist across commands and specific to the environment, but should also be a flag that overrides it, because the use of it could be scoped to a single execution of a command (which would facilitate testing.)

As an example, I looked into what the AWS cli does. AWS has a [set of command line options](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-options.html) that apply to all CLI commands regardless of the service being called. These are options that have defaults and are completely optional. Most notably are `--profile` which allows you to use multiple **Config** profiles, and `--endpoint-url`.

For **unit testing**, if we can decouple testing the behavior (when passing in `x`, the cli does `y`) from testing the API (`x` is passed in using a flag/config/env and results in calling foo with `x`), we could get good coverage and confidence.

For **integration tests**, we should be testing exactly what the user would run, so we should ensure that setting an environment variable does what we expect it to do. That might mean grouping together some tests using the same `MockServer`?
